src/NUnitFramework/framework/Internal/Execution/ParallelWorkItemDispatcher.cs

```diff
@@ -221,6 +226,15 @@ private void Dispatch(WorkItem work, ParallelExecutionStrategy strategy)
         {
             log.Debug("Using {0} strategy for {1}", strategy, work.Name);
 
+            // Currently, we only track CompositeWorkItems - this could be expanded
+            var composite = work as CompositeWorkItem;
+            if (composite != null)
+                lock (_activeWorkItems)
+                {
+                    _activeWorkItems.Add(composite);
+                    composite.Completed += OnWorkItemCompletion;
```

Comment:
Would it be possible to run this line before taking the lock?

Reply:
I believe it is, but __only__ because we know how the containing method is being called. And it could eventually be called incorrectly. If that happened, we might end up with a race condition where the completion event that removes the item got called before the item was added. This way seemed safer.

Reply:
That sounds best.